### PR TITLE
feat: support downloading certificates

### LINKED ISSUE #47: cli supports downloading certificates
@bjwswang How about this feature?

### PR DESCRIPTION
**Pull request by dayuy**
Fix #47

### CODE REVIEW

File: pkg/depository/types.go [hunk 1]
```diff
@@ -35,7 +34,7 @@ type Depository struct {
 	ContentName      string `json:"contentName" pg:"contentName"`
 	ContentID        string `json:"contentID" pg:"contentID"`
 	ContentType      string `json:"contentType" pg:"contentType"`
-	TrustedTimestamp string `json:"trustedTimestamp" pg:"trustedTimestamp"`
+	TrustedTimestamp int64  `json:"trustedTimestamp" pg:"trustedTimestamp"`
```
[the remark this answers is not in the document]
Comment (smoky8) [reply pkg/depository/types.go:37]: No. The table have been updated.

File: pkg/certificate/get.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2023 The Bestchains Authors.
```
Comment (bjwswang) [inline pkg/certificate/get.go:2]: Should we move this unduer `pkg/depository`  as this only works for `depository service`

Comment (0xff-dev) [reply pkg/certificate/get.go:2]: Agree.

File: pkg/certificate/get.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+)
+
+func NewDepositoryCertificateGetCmd(option common.Options) *cobra.Command {
+	var kid string
+
+	cmd := &cobra.Command{
+		Use:   "certificate",
+		Short: "Download the depository certificates",
+		Run: func(cmd *cobra.Command, args []string) {
+			host := viper.GetString("saas.depository.server")
```
Comment (0xff-dev) [inline pkg/certificate/get.go:38]: In depository, the `host` parameter is defined and bound to `saas.depository.server`, currently the `certificate command` does not define the `host` parameter, so the host obtained here is empty.

I think we should put the host parameter under the root cmd, And remove the `host` parameter from depository's `create` and `get` commands.

Comment (0xff-dev) [reply pkg/certificate/get.go:38]:
```go
// cmd/bc-cli/main.go
func NewCmd() *cobra.Command {
   ...
    cmd.PersistentFlags().StringP("host", "http://localhost:9999", "host URL of depository server")
    _ = viper.BindPFlag("saas.depository.server", cmd.PersistentFlags().Lookup("host"))
}
```

**Comment by 0xff-dev** [comment]
The running result looks good.
```shell
root@macbookpro:~/workspace/bc-cli/cmd/bc-cli# ./main get certificate --host='https://bc-saas.172.22.96.209.nip.io' --kid=18b682430b2ee5cdcb17c0ea678a52223056d256
certificate 18b682430b2ee5cdcb17c0ea678a52223056d256.pdf downloaded
```

File: pkg/certificate/get.go (hunk 2 above)
Comment (bjwswang) [reply pkg/certificate/get.go:2]: @dayuy

Comment (0xff-dev) [reply pkg/certificate/get.go:2]: > @dayuy

Any suggestions for command format?
```shell
#? 
bc-cli get depository certificate --kind
```

File: pkg/certificate/get.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+			if host == "" {
+				fmt.Fprintln(option.ErrOut, "no host provided")
+				return
+			}
+			u := fmt.Sprintf("%s%s", host, fmt.Sprintf(common.DepositoryCertificate, kid))
+			headers := map[string]string{
+				"Content-Type":        "application/octet-stream",
+				"Content-Disposition": fmt.Sprintf("attachment; filename=%s.pdf", kid),
```
Comment (0xff-dev) [inline pkg/certificate/get.go:46]: Add `Authorization` request header.

Comment (dayuy) [reply pkg/certificate/get.go:46]: The `NewCmd` function will always add  the `Authorization` request header.

**Comment by dayuy** [comment]
```shell
➜  bc-cli git:(feat-7) ✗ ./main get depository cd97ee2faf890a3b3a5527ee9571a735aa3e4b0c 5635ab1159528b88a099cccf81a3f2f0bb6f438f --certificate
certificate cd97ee2faf890a3b3a5527ee9571a735aa3e4b0c.pdf downloaded
certificate 5635ab1159528b88a099cccf81a3f2f0bb6f438f.pdf downloaded
``` 

image

**Comment by 0xff-dev** [comment]
> ```shell
> ➜  bc-cli git:(feat-7) ✗ ./main get depository cd97ee2faf890a3b3a5527ee9571a735aa3e4b0c 5635ab1159528b88a099cccf81a3f2f0bb6f438f --certificate
> certificate cd97ee2faf890a3b3a5527ee9571a735aa3e4b0c.pdf downloaded
> certificate 5635ab1159528b88a099cccf81a3f2f0bb6f438f.pdf downloaded
> ```
> 
> image

@bjwswang  The `depository hash` contain a `.png` suffix.

**Comment by dayuy** [comment]
> 
> @bjwswang The `depository hash` contain a `.png` suffix.

the `.png` suffix indicates that the depository file is an image stored in MinIO

image

**Comment by bjwswang** [comment]
@dayuy how did you get the png hash ?we can only keep the first 20bytes of the hash

**Comment by bjwswang** [comment]
> @dayuy how did you get the png hash ?we can only keep the first 20bytes of the hash . Please refer to [kid calculation](https://github.com/bestchains/bestchains-contracts/blob/main/contracts/depository/depository.go#L401)

Besides now the certificate allows cell size defintion in each location. See https://github.com/bestchains/bc-saas/pull/52

File: pkg/depository/get.go [hunk 5]
```diff
@@ -66,6 +67,40 @@ func NewGetDepositoryCmd(option common.Options) *cobra.Command {
...
 				return fmt.Errorf("no host provided")
 			}
+			certificate, _ := cmd.Flags().GetBool("certificate")
+			if certificate {
+				if len(args) == 0 {
+					return fmt.Errorf("no kid provided")
+				}
+				for _, kid := range args {
+					u := fmt.Sprintf("%s%s", host, fmt.Sprintf(common.DepositoryCertificate, kid))
+					headers := map[string]string{
```
Comment (0xff-dev) [inline pkg/depository/get.go:77]: There is no need to add a request header.

File: pkg/depository/get.go [hunk 6]
```diff
@@ -66,6 +67,40 @@ func NewGetDepositoryCmd(option common.Options) *cobra.Command {
 			if host == "" {
 				return fmt.Errorf("no host provided")
 			}
+			certificate, _ := cmd.Flags().GetBool("certificate")
```
Comment (bjwswang) [inline pkg/depository/get.go:70]: add a new flag `certificate-language` which supports :
- "CN" 
- "ENG"

And this argument should be passed into http request with "?language="xxx""

**Comment by dayuy** [comment]
> @dayuy how did you get the png hash ?we can only keep the first 20bytes of the hash . Please refer to [kid calculation](https://github.com/bestchains/bestchains-contracts/blob/main/contracts/depository/depository.go#L401)

File: pkg/depository/get.go (hunk 6 above)
Comment (dayuy) [reply pkg/depository/get.go:70]: is it `?style=xxx` or `?language=ENG`?
the `bc-saas` displays `language := ctx.Query("language")`

**Review by 0xff-dev — APPROVED** [review]

**Review by Abirdcfly — APPROVED** [review]

File: pkg/depository/get.go (hunk 6 above)
Comment (bjwswang) [reply pkg/depository/get.go:70]: use `language` for now.

**Review by bjwswang — APPROVED** [review]

**Pull request merged**